packages/server/src/unstable-core-do-not-import/stream/stream.ts

```diff
@@ -0,0 +1,575 @@
+const CHUNK_VALUE_TYPE_ASYNC_ITERABLE = 1;
+type CHUNK_VALUE_TYPE_ASYNC_ITERABLE = typeof CHUNK_VALUE_TYPE_ASYNC_ITERABLE;
+
+const PROMISE_STATUS_FULFILLED = 0;
+type PROMISE_STATUS_FULFILLED = typeof PROMISE_STATUS_FULFILLED;
+const PROMISE_STATUS_REJECTED = 1;
+type PROMISE_STATUS_REJECTED = typeof PROMISE_STATUS_REJECTED;
+
+const ASYNC_ITERABLE_STATUS_DONE = 0;
+type ASYNC_ITERABLE_STATUS_DONE = typeof ASYNC_ITERABLE_STATUS_DONE;
+const ASYNC_ITERABLE_STATUS_VALUE = 1;
+type ASYNC_ITERABLE_STATUS_VALUE = typeof ASYNC_ITERABLE_STATUS_VALUE;
+const ASYNC_ITERABLE_STATUS_ERROR = 2;
+type ASYNC_ITERABLE_STATUS_ERROR = typeof ASYNC_ITERABLE_STATUS_ERROR;
+
+type ChunkDefinitionKey =
+  // root should be replaced
+  | null
+  // at array path
+  | number
+  // at key path
+  | string;
+
+type ChunkIndex = number & { __chunkIndex: true };
+type ChunkValueType =
+  | CHUNK_VALUE_TYPE_PROMISE
+  | CHUNK_VALUE_TYPE_ASYNC_ITERABLE;
+type ChunkDefinition = [
+  key: ChunkDefinitionKey,
+  type: ChunkValueType,
+  chunkId: ChunkIndex,
+];
+type HydratedValue = [
+  // data
+  [unknown],
+  // chunk descriptions
+  ...ChunkDefinition[],
+];
+
+type Head = Record<string, HydratedValue>;
+type PromiseChunk =
+  | [
+      chunkIndex: ChunkIndex,
+      status: PROMISE_STATUS_FULFILLED,
+      value: HydratedValue,
+    ]
+  | [
+      chunkIndex: ChunkIndex,
+      status: PROMISE_STATUS_REJECTED,
+      // do we want to serialize errors?
+      // , error?: unknown
+    ];
+type IterableChunk =
+  | [chunkIndex: ChunkIndex, status: ASYNC_ITERABLE_STATUS_DONE]
+  | [
+      chunkIndex: ChunkIndex,
+      status: ASYNC_ITERABLE_STATUS_VALUE,
+      value: HydratedValue,
+    ]
+  | [
+      chunkIndex: ChunkIndex,
+      status: ASYNC_ITERABLE_STATUS_ERROR,
+      // do we want to serialize errors?
+      // , error?: unknown
+    ];
+type ChunkData = PromiseChunk | IterableChunk;
+type PlaceholderValue = 0 & { __placeholder: true };
+export function isPromise(value: unknown): value is Promise<unknown> {
+  return isObject(value) && typeof (value as any).then === 'function';
+}
+
+type Serialize = (value: any) => any;
+type Deserialize = (value: any) => any;
+
+export type ProducerOnError = (opts: {
+  error: unknown;
+  path: (string | number)[];
+}) => void;
+export interface ProducerOptions {
+  serialize?: Serialize;
+  data: Record<number, unknown>;
+  onError?: ProducerOnError;
+  maxDepth?: number;
+}
+
+class MaxDepthError extends Error {
+  constructor(public path: (string | number)[]) {
+    super('Max depth reached at path: ' + path.join('.'));
+  }
+}
+
+export function createBatchStreamProducer(opts: ProducerOptions) {
+  const { data } = opts;
+  let counter = 0 as ChunkIndex;
+  const placeholder = 0 as PlaceholderValue;
+
+  const [stream, controller] = createReadableStream<ChunkData>();
+  const pending = new Set<ChunkIndex>();
+  function maybeClose() {
+    if (pending.size === 0) {
+      controller.close();
+    }
```

Comment:
I'm not sure if there's a race condition in which this could be called prematurely